[MRG+1] UnaryEncoder to encode ordinal features into unary levels

arjunjauhari opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue
<!-- Example: Fixes #1234 -->

Resolves #8628 
Resolves #3336

#### What does this implement/fix? Explain your changes.
This implements OrdinalEncoder, which is a more informative encoding than one-hot for ordinal features. Implemented a new class OrdinalEncoder whose interface is same as OneHotEncoder class.

#### Any other comments?
Logic: For k values 0, ..., k - 1 of the ordinal feature x, this creates k - 1 binary features such that the ith is active if x > i (for i = 0, ... k - 1)

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

**Working Example**
```py
>>> from sklearn.preprocessing import OrdinalEncoder
>>> enc = OrdinalEncoder()
>>> enc.fit([[0, 0, 3], [1, 1, 0], [0, 2, 1], [1, 0, 2]])  
OrdinalEncoder(dtype=<type 'numpy.float64'>, handle_unknown='error',
        n_values='auto', ordinal_features='all', sparse=True)
>>> 
>>> enc.n_values_
array([2, 3, 4])
>>> 
>>> enc.feature_indices_
array([0, 1, 3, 6])
>>> 
>>> enc.active_features_
array([0, 1, 2, 3, 4, 5])
>>>
>>> enc.transform([[0, 1, 1]]).toarray()
array([[ 0.,  1.,  0.,  1.,  0.,  0.]])
```
**Another One**
```
>>> OrdinalEncoder(3).fit_transform([[0], [1], [2]]).toarray()
array([[ 0.,  0.],
       [ 1.,  0.],
       [ 1.,  1.]])
```
**A to-do list**
- [x] Converge on a final name for this feature. Current suggestions: `UnaryEncoder[+2]`, `OrdinalEncoder[0]`
- [x] Fix the nosetests errors
- [x] Write examples
- [x] Write unit-test cases
- [x] Any documentation work

arjunjauhari commented:
@jnothman @jmschrei 
Detailed comment for #8628 

**Working Example**
```
>>> from sklearn.preprocessing import OrdinalEncoder
>>> enc = OrdinalEncoder()
>>> enc.fit([[0, 0, 3], [1, 1, 0], [0, 2, 1], [1, 0, 2]])  
OrdinalEncoder(dtype=<type 'numpy.float64'>, handle_unknown='error',
        n_values='auto', ordinal_features='all', sparse=True)
>>> 
>>> enc.n_values_
array([2, 3, 4])
>>> 
>>> enc.feature_indices_
array([0, 1, 3, 6])
>>> 
>>> enc.active_features_
array([0, 1, 2, 3, 4, 5])
>>>
>>> enc.transform([[0, 1, 1]]).toarray()
array([[ 0.,  1.,  0.,  1.,  0.,  0.]])
```
**Another One**
```
>>> OrdinalEncoder(3).fit_transform([[0], [1], [2]]).toarray()
array([[ 0.,  0.],
       [ 1.,  0.],
       [ 1.,  1.]])
```
**A to-do list**
- [ ] Fix the nosetests errors
- [ ] Write examples
- [ ] Write unit-test cases
- [ ] Any documentation work

**Questions for reviewers**
- How to we handle when all the features in training data has just one value. Currently it return empty array. Is there a better way to handle this? For example,
```
>>> enc = OrdinalEncoder()
>>> enc.fit_transform([[0,0],[0,0],[0,0]]).toarray()
array([], shape=(3, 0), dtype=float64)
```

**Need help with these 3 errors which i get while running nosetests after build**
```
======================================================================
ERROR: /Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/tests/test_common.py.test_non_meta_estimators:check_transformer_general(OrdinalEncoder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arjunjauhari/venvs/py2gen/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/testing.py", line 741, in __call__
    return self.check(*args, **kwargs)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/testing.py", line 292, in wrapper
    return fn(*args, **kwargs)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/estimator_checks.py", line 662, in check_transformer_general
    _check_transformer(name, Transformer, X, y)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/estimator_checks.py", line 744, in _check_transformer
    % Transformer)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/numpy/testing/utils.py", line 811, in assert_array_almost_equal
    header=('Arrays are not almost equal to %d decimals' % decimal))
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/numpy/testing/utils.py", line 633, in assert_array_compare
    reduced = val.ravel()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/scipy/sparse/base.py", line 440, in __getattr__
    raise AttributeError(attr + " not found")
AttributeError: ravel not found

======================================================================
FAIL: sklearn.feature_extraction.tests.test_image.test_connect_regions_with_grid
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arjunjauhari/venvs/py2gen/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/feature_extraction/tests/test_image.py", line 86, in test_connect_regions_with_grid
    assert_equal(ndimage.label(mask)[1], connected_components(graph)[0])
AssertionError: 200 != 199
    '200 != 199' = '%s != %s' % (safe_repr(200), safe_repr(199))
    '200 != 199' = self._formatMessage('200 != 199', '200 != 199')
>>  raise self.failureException('200 != 199')
    

======================================================================
FAIL: /Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/tests/test_common.py.test_all_estimators:check_parameters_default_constructible(OrdinalEncoder)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/arjunjauhari/venvs/py2gen/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/testing.py", line 741, in __call__
    return self.check(*args, **kwargs)
  File "/Users/arjunjauhari/Desktop/personal/work/scikit-learn/sklearn/utils/estimator_checks.py", line 1598, in check_parameters_default_constructible
    np.float64, types.FunctionType, Memory])
AssertionError: <type 'type'> not found in [<type 'str'>, <type 'int'>, <type 'float'>, <type 'bool'>, <type 'tuple'>, <type 'NoneType'>, <type 'numpy.float64'>, <type 'function'>, <class 'sklearn.externals.joblib.memory.Memory'>]
    """Fail immediately, with the given message."""
>>  raise self.failureException("<type 'type'> not found in [<type 'str'>, <type 'int'>, <type 'float'>, <type 'bool'>, <type 'tuple'>, <type 'NoneType'>, <type 'numpy.float64'>, <type 'function'>, <class 'sklearn.externals.joblib.memory.Memory'>]")
```    
_Thanks_

raghavrv commented:
Thanks for the PR... I think `UnaryEncoder` is more apt? As these are infact [unary numbers](https://en.wikipedia.org/wiki/Unary_numeral_system)...

raghavrv commented:
Also what does `active_features_` denote?

> How to we handle when all the features in training data has just one value

Just return `[0] * n_features`? Returning an empty matrix is not good as it can lead to errors when used in a pipeline...

raghavrv commented:
I've updated the PR description based on your subsequent comment. Hope you don't mind...

raghavrv commented on sklearn/preprocessing/data.py:
```diff
@@ -1868,7 +1868,7 @@ def _transform(self, X):
         if np.any(~mask):
             if self.handle_unknown not in ['error', 'ignore']:
                 raise ValueError("handle_unknown should be either error or "
```
Can you also make it `either 'error' or 'ignore'`?

raghavrv replied:
(surround with single quotes)

arjunjauhari replied:
Done. Also updated the similar error message for OneHotEncoder.

raghavrv commented on sklearn/preprocessing/data.py:
```diff
@@ -1901,3 +1901,241 @@ def transform(self, X):
+        return self
+
+    def _fit_transform(self, X):
+        """Assumes X contains only oridinal features."""
```
ordinal

arjunjauhari replied:
Fixed

raghavrv commented on sklearn/preprocessing/data.py:
```diff
@@ -1901,3 +1901,241 @@ def transform(self, X):
+            except (ValueError, TypeError):
+                raise TypeError("Wrong type for parameter `n_values`. Expected"
+                                " 'auto', int or array of ints, got %r"
+                                % type(X))
```
I think you can just print it, rather than printing the type...

arjunjauhari replied:
Done

raghavrv reviewed (dismissed):
A few starting comments...

arjunjauhari commented:
Thanks for reviewing Venkat.
> I think UnaryEncoder is more apt? As these are infact unary numbers...

On naming, I feel `UnaryEncoder` makes sense at some level. But want to highlight that even though individual feature value gets encoded as a Unary system, still the final output is binary. 
> Also what does active_features_ denote?

Thanks for bringing this up, in current implementation it doesn't denote anything of use because all the features in output will always be active. I will remove it. In my first implementation, it had some relevance.
> Just return [0] * n_features? Returning an empty matrix is not good as it can lead to errors when used in a pipeline...

But then for this input `>>> enc.fit_transform([[0,0],[0,0],[0,0]])` we will get output `[0, 0]`. I don't think that's what we want. Let me know if I am missing something. 
How about we do one of these: 
- Either we return the X itself back
- or we can return an array with shape (n_samples, 1) filled with zeros. For about input it will be `[[0],[0],[0]]`

> I've updated the PR description based on your subsequent comment. Hope you don't mind...

Actually thanks, looks cleaner.

arjunjauhari commented:
Also this encoder can give same output for two different inputs. For example,
```
>>> enc.fit_transform([[0,0],[0,0],[0,1]]).toarray()
array([[ 0.],
       [ 0.],
       [ 1.]])
>>> enc.fit_transform([[0],[0],[1]]).toarray()
array([[ 0.],
       [ 0.],
       [ 1.]])
```
But i think thats okay as anyway first feature is not holding anything meaningful. Let me know if you feel some concern around it.

arjunjauhari committed:
Rebase and commit the OrdinalEncoder implementation

raghavrv commented:
> still the final output is binary

I think I'd call it more of a "concatenation of unary number sequences" rather than binary... I am strongly inclining towards `UnaryEncoder`... Opinions may vary though :)

> Also this encoder can give same output for two different inputs. For example,

I think it should actually be -

```
>>> enc.fit_transform([[0, 0], [0, 0], [0, 1]]).toarray()
array([[ 0, 0],
       [ 0, 0],
       [ 0, 1]])
>>> enc.fit_transform([[0],[0],[1]]).toarray()
array([[ 0],
       [ 0],
       [ 1]])
```

raghavrv commented:
Such that `n_new_features >= n_features`...

jnothman commented:
I think unary is less ambiguous than ordinal

On 29 Mar 2017 4:17 am, "(Venkat) Raghav (Rajagopalan)" <
notifications@github.com> wrote:

> Such that n_new_features >= n_features...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8652#issuecomment-289840959>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69zwPjKO-B1rI_I-C4ksuZWVh--qks5rqUCpgaJpZM4MpxOA>
> .
>

arjunjauhari commented:
> Such that n_new_features >= n_features...

is this a requirement for all encoders to satisfy? why?
As I don't see a problem with `n_new_features < n_features` since its a lossless dimensionality reduction it this case.

jnothman commented:
I'm okay with discarding input features that are constant. It's fine to
generally output k features for a feature with k values.

On 29 March 2017 at 11:51, Arjun Jauhari <notifications@github.com> wrote:

> Such that n_new_features >= n_features...
>
> is this a requirement for all encoders to satisfy? why?
> As I don't see a problem with n_new_features < n_features since its a
> lossless dimensionality reduction it this case.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8652#issuecomment-289949002>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz67XRFz7ElQ8bRaERhnA8-S7UvubEks5rqar0gaJpZM4MpxOA>
> .
>

arjunjauhari commented:
> It's fine to generally output k features for a feature with k values.

As of now I output, 
- k-1 features for a input feature with k values (so it ends up in discarding input feature with just 1 value)

I can also do, no output features for input feature that are constant(as @jnothman suggested).

But all this makes `n_out_features < n_features(in)` a real possibility. The question is if that's acceptable or not?

jnothman commented:
having fewer output features is fine imo

On 30 Mar 2017 2:29 am, "Arjun Jauhari" <notifications@github.com> wrote:

> It's fine to generally output k features for a feature with k values.
>
> As of now I output,
>
>    - k-1 features for a input feature with k values (so it ends up in
>    discarding input feature with just 1 value)
>
> I can also do, no output features for input feature that are constant(as
> @jnothman <https://github.com/jnothman> suggested).
>
> But all this makes n_out_features < n_features(in) a real possibility.
> The question is if that's acceptable or not?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8652#issuecomment-290127082>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63JFUxKXFqwzRXXUH4hv6mTS4hdMks5rqnjAgaJpZM4MpxOA>
> .
>

jnothman commented:
@arjunjauhari, are you still working on this?

More than anything else, this needs tests in `sklearn/preprocessing/tests/test_data.py`.

Here's something I had lying around which may apply, to get started:

```python
def test_ordinal_encoder():
    X = np.arange(-1, 5).reshape(-1, 1)
    est = OrdinalEncoder(4)
    Xt = est.fit_transform(X)
    assert_array_equal(Xt, [[0, 0, 0],   # -1
                            [0, 0, 0],   # 0
                            [1, 0, 0],   # 2
                            [1, 1, 0],   # 3
                            [1, 1, 1],   # 4
                            [1, 1, 1]])  # 5
    Xt2 = est.transform(X)
    assert_array_equal(Xt2, Xt)

    # smaller n_values leads to horizontally truncated output
    Xt3 = OrdinalEncoder(3).fit_transform(X)
    assert_array_equal(Xt3, np.array(Xt2)[:, :-1])

    # multiple input features stacks output
    rng = np.random.RandomState(0)
    X_multi = rng.randint(3, size=(10, 3))
    X_multi_t = OrdinalEncoder(3).fit_transform(X_multi)
    assert_equal(X_multi_t.shape, (10, 3 * 2))
    expected = np.hstack([OrdinalEncoder(3).fit_transform(X_multi[:, i:i + 1])
                          for i in range(X_multi.shape[1])])
    assert_array_equal(expected, X_multi_t)
```

jnothman commented:
After 20 days of silence from @arjunjauhari, I'm marking this as Need Contributor. @arjunjauhari, let me know if you would rather keep working on it.

ruxandraburtica commented:
@jnothman I will continue working on this, if that's ok

jnothman commented:
You're welcome to open a PR continuing the work (or start again) as far as I'm concerned.

Note that we're currently trying to release version 0.19, so if you don't get quicky responses, you should ping after a few weeks

arjunjauhari commented:
@ruxandraburtica I am planning to continue this work and take PR to completion. I hope you have not started yet. Sorry for any inconvenience. 
@jnothman Sorry for my absence in last month. I will complete this PR on priority. I hope that's okay.

jnothman commented:
that's fine. no great hurry, but it is hard to know when there's no
response what the status is

On 23 Jun 2017 6:51 pm, "Arjun Jauhari" <notifications@github.com> wrote:

> @ruxandraburtica <https://github.com/ruxandraburtica> I am planning to
> continue this work and take PR to completion. I hope you have not started
> yet. Sorry for any inconvenience.
> @jnothman <https://github.com/jnothman> Sorry for my absence in last
> month. I will complete this PR on priority. I hope that's okay.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8652#issuecomment-310610450>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz62CfF-IERtTXDuIA3MkLqLxiu2Mrks5sG3yegaJpZM4MpxOA>
> .
>

arjunjauhari commented:
@jnothman Thanks, I will continue working.

ruxandraburtica commented:
@jnothman @arjunjauhari actually I started working on this as well, I didn't open a pull request because there were a few more tests I wanted to add. I'll be at a computer in about 2 hours, and create the PR.

arjunjauhari committed:
Updating name to UnaryEncoder and adding single quote in error string

ruxandraburtica commented:
@jnothman @arjunjauhari I've opened a PR here: https://github.com/scikit-learn/scikit-learn/pull/9216

It still needs some tests, but I wanted to know if I should continue working on it.

arjunjauhari commented:
@ruxandraburtica I am happy to merge your work with mine, if that's okay to you. That way none of our work go unused.

jnothman commented:
@ruxandraburtica, You should be able to open a pull request to @arjunjauhari's branch, if that helps merge your work. Sorry for the miscommunication

ruxandraburtica commented:
@arjunjauhari @jnothman yes, I will do that today

ruxandraburtica committed:
Merged changes from #9216

ruxandraburtica commented:
@arjunjauhari is it possible that you add me as a contributor to your fork of `scikit-learn`? Because I'm trying to create a PR to your repository, but not all forks are visible: 
<img width="1130" alt="screen shot 2017-06-25 at 2 41 49 pm" src="https://user-images.githubusercontent.com/797922/27515795-bf48ef2a-59b4-11e7-9fbd-dcdb412bf4bc.png">

arjunjauhari commented:
@ruxandraburtica Added.

arjunjauhari commented:
@ruxandraburtica Thanks for merging your changes to my branch. I see that you added the `active_features_` attribute to the UnaryEncoder class but as discussed above in this thread we don't need it as all the features in output will be active anyway. So I will go ahead and remove it. Let me know if you have any other thoughts.

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -497,6 +497,53 @@ features, one has to explicitly set ``n_values``. For example,
+Encoding ordinal features
+=============================
+Often categorical features have a clear ordering. For example a person could have features
+``["short", "tall"]``,
```
Format with bullet points

arjunjauhari replied:
done

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -497,6 +497,53 @@ features, one has to explicitly set ``n_values``. For example,
+    >>> enc.transform([[0, 1, 1]]).toarray()
+    array([[ 0.,  1.,  0.,  1.,  0.,  0.]])
+
+By default, how many values each feature can take is inferred automatically from the dataset.
```
please limit your lines to under 80 characters

arjunjauhari replied:
done

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    the values taken on by ordinal (discrete) features. The output will be
+    a sparse matrix where each column corresponds to one possible value of one
+    feature. It is assumed that input features take on values in the range
+    [0, n_values).
```
While this interval notation is well-known, I think "from 0 to n_values - 1" is much clearer.

arjunjauhari replied:
done

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+        Will return sparse matrix if set True else will return an array.
+
+    handle_unknown : str, 'error' or 'ignore'
+        Whether to raise an error or ignore if a unknown ordinal feature is
```
I think we should always raise an error for values < 0. Please clarify this in the description (and implement and test if not the current behaviour).

arjunjauhari replied:
Already in place

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    dtype : number type, default=np.float
+        Desired dtype of output.
+
+    sparse : boolean, default=True
```
I think sparse should be false by default. If the data is uniformly distributed, then sparse=True will lead to a matrix with 50% sparsity which is not enough to make a sparse representation efficient. Less efficient still if `ordinal_features` is set.

arjunjauhari replied:
changed default to False

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    >>> from sklearn.preprocessing import UnaryEncoder
+    >>> enc = UnaryEncoder()
+    >>> enc.fit([[0, 0, 3], [1, 1, 0], [0, 2, 1], \
+[1, 0, 2]])  # doctest: +ELLIPSIS
```
This is better as:
```python
    >>> enc.fit([[0, 0, 3], [1, 1, 0], [0, 2, 1],
    ...          [1, 0, 2]])  # doctest: +ELLIPSIS
```

jnothman replied:
Or even:
```python
    >>> enc.fit([[0, 0, 3],
    ...          [1, 1, 0],
    ...          [0, 2, 1],
    ...          [1, 0, 2]])  # doctest: +ELLIPSIS
```

arjunjauhari replied:
done

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+        Indices to feature ranges.
+        Feature ``i`` in the original data is mapped to features
+        from ``feature_indices_[i]`` to ``feature_indices_[i+1]``
+        (and then potentially masked by `active_features_` afterwards)
```
should use double backticks, not single...

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+
+    Attributes
+    ----------
+    active_features_ : array
```
Did we decide against having this?

arjunjauhari replied:
yes, removed it

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+    array([0, 1, 3, 6])
+    >>> enc.active_features_
+    array([0, 1, 2, 3, 4, 5])
+    >>> enc.transform([[0, 1, 1]]).toarray()
```
It's not very useful to transform only 1s. Use a 2 in there, please.

arjunjauhari replied:
done

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+
+    See also
+    --------
+    sklearn.feature_extraction.DictVectorizer : performs a Ordinal encoding of
```
These are not all relevant. Just point to `OneHotEncoder`.

arjunjauhari replied:
done

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2472,3 +2475,239 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+                raise TypeError("Wrong type for parameter `n_values`. Expected"
+                                " 'auto', int or array of ints, got %r"
+                                % self.n_values)
+            if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
```
this if statement is never reached in current tests.

arjunjauhari replied:
Added a new test to cover it

jnothman reviewed (commented):
I'm not sure if @ruxandraburtica or @arjunjauhari is currently running the show here. Thanks to whomever can continue the work. Preprocessing is getting a lot of attention this release, so it would be good to get this in too.

Firstly, please merge in master and resolve conflicts. I've only reviewed API so far and a couple of other small things, not implementation nor tests.

arjunjauhari commented:
@jnothman Thanks for reviewing and bringing this up to notice. Let me take this to completion. I will take a look now.

arjunjauhari commented:
@jnothman @raghavrv I have rebased(resolved conflicts) ordinal-encoder feature branch onto master branch. Should i go ahead and force push to update the ordinal-encoder remote feature branch?

jnothman commented:
you may. I prefer git merge master over git rebase master (it's easier and
cleaner and force push is unnecessary) but if the rebase is done, go ahead.

arjunjauhari committed:
Removing active_features_ attribute from UnaryEncoder as it is not needed

arjunjauhari committed:
Limiting the lines in documentation to less that 80 chars

arjunjauhari committed:
Updated documentation. Changed the default value of sparse parameter to False

arjunjauhari committed:
Updated test cases to accomodate change in default value of sparse parameter. Also added a new test case test_unary_encoder_n_values_array

arjunjauhari commented:
@jnothman @raghavrv Updated PR to take care of all the comments. I think its ready for detailed review. Please assign reviewers.

jnothman commented:
You have test failures...

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,218 @@ def inverse_transform(self, X):
+
+        # We use only those ordinal features of X that are known using fit.
+        # i.e lesser than n_values_ using mask.
+        # This means, if self.handle_unknown is "ignore", the row_indices and
```
I think I had a different idea of `handle_unknown='ignore'` semantics, and maybe we need to change the name:
* if x < 0: raise
* if 0 <= x < n_values: set x 1s in the vector
* if x >= n_values: set n_values 1s in the vector

i.e. in this case, all should be full. I think this makes sense for count data, for instance.

Maybe we should rename it to `handle_unknown='fill`' or `handle_greater={'error', 'fill'}`. Or maybe it could be called 'clip' instead of 'fill'.

arjunjauhari committed:
Commit to accomodate all the requested changes
  1. Fixed test failures
  2. Updated docs
  3. UnaryEncoder: Changed handle_unknown to handle_greater

Updated docs

UnaryEncoder: Changed handle_unknown to handle_greater

arjunjauhari committed:
Fixing DocTestFailure

arjunjauhari commented:
@jnothman @raghavrv Updated PR with requested changes.
Also, I saw that in `doc/modules/preprocessing.rst`, documentation for `preprocessing_categorical_features` has more than 80 chars at several places. Should I go ahead and clean it up in a separate PR?

jnothman commented:
Don't worry too much about cleaning stylistic things in existing code and
docs. We find such changes more often disruptive than helpful. I'll check
in on this later

jnothman commented on doc/modules/preprocessing.rst:
```diff
@@ -559,6 +559,12 @@ implemented in :class:`UnaryEncoder`.  This estimator transforms each
 ordinal feature with ``m`` possible values into ``m - 1`` binary features,
 where the ith feature is active if x > i (for i = 0, ... k - 1).
 
+**NOTE**: This encoding is likely to help when used with linear models and
```
notes can be formatted with `.. note::`

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2919,9 +2923,10 @@ class UnaryEncoder(BaseEstimator, TransformerMixin):
-        Whether to raise an error or ignore if a unknown ordinal feature is
-        present during transform.
+    handle_greater : str, 'error' or 'clip'
+        Whether to raise an error or clip if a greater ordinal feature value is
```
"Whether to raise an error or clip if an ordinal feature >= n_values is passed in" is more precise, I think.

I think the same behaviour should happen in fit if `n_values` is specified.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -3046,28 +3064,28 @@ def _transform(self, X):
 
-        column_start = np.tile(indices[:-1], n_samples)[mask]
-        column_end = (X + indices[:-1]).ravel()[mask]
+        X_ceil = np.where(mask.reshape(X.shape), X, self.n_values_ - 1)
```
Now is the transformation logic identical in the fit and transform cases? Are we now able to define fit_transform as just fit + transform?

jnothman replied:
What I mean is we should be able to refactor this.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
+    # Test that encoder raises error for greater features.
+    encoder = UnaryEncoder(handle_greater='error')
+    encoder.fit(X)
+    assert_raises(ValueError, encoder.transform, y)
```
We should check these behaviours for fit_transform too... And I wonder about whether we should raise an error if we set n_values explicitly and a `fit` feature exceeds it, even if `fit_transform` is not called then.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,236 @@ def inverse_transform(self, X):
+            All feature values should be non-negative otherwise will raise a
+            ValueError.
+        """
+        self.fit_transform(X)
```
I don't think we should be doing unnecessary transforming here.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
+
+    # Raise error if handle_greater is neither ignore or error.
+    encoder = UnaryEncoder(handle_greater='42')
+    encoder.fit(X)
```
All parameter validation should happen in fit. This is where we should expect the error to be raised.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
+    X = _generate_random_features_matrix(n_values, size)
+    encoder.fit(X)
+
+    # test that an error is raised when different shape
```
This should be covered by common tests.

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
+    assert_array_equal(
+        enc.feature_indices_,
+        np.arange(0, unary_n_values * len(X[0]) + 1, unary_n_values)
+    )
```
Please test that `n_values_` is correct

jnothman commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,175 @@ def test_quantile_transform_valid_axis():
+
+    enc = UnaryEncoder(n_values=[])
+    X = _generate_random_features_matrix(n_features, size)
+    assert_raises(ValueError, enc.fit_transform, X)
```
Please test that `n_values_` is correct

jnothman reviewed (commented):
Tests are looking pretty good!

arjunjauhari committed:
Refactoring the code. Now fit_transform in equivalent to fit +
transform.
Parameters checking being done in fit method. Plus, new test cases.

Updated docs

Updating implementation of UnaryEncoder

checking handle_greater parameter value in fit

New test cases

arjunjauhari commented:
@jnothman Refactored code a bit. Now actually fit_transform is equivalent to fit + transform. Also added few new test cases.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,229 @@ def inverse_transform(self, X):
+
+        return X
+
+    def fit_transform(self, X, y=None):
```
You can remove this and rely on the implementation in `TransformerMixin`...

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -3016,24 +3019,18 @@ def _fit_transform(self, X):
-        out = sparse.coo_matrix((data, (row_indices, column_indices)),
-                                shape=(n_samples, indices[-1]),
-                                dtype=self.dtype).tocsr()
+        mask = (X < self.n_values_).ravel()
```
use >= and you don't need subsequent `~`s

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -3016,24 +3019,18 @@ def _fit_transform(self, X):
+        mask = (X < self.n_values_).ravel()
+        if np.any(~mask):
+            if self.handle_greater == 'error':
+                raise ValueError("unknown ordinal feature present %s "
```
how about "handle_greater='error' but feature value %d > n_values_[%d] = %d" and just give a single example...?

arjunjauhari replied:
By give single example do you mean, if there are multiple feature value which are failing, just raise the error printing any one of them?

jnothman replied:
I think that's at least as useful as printing all of them. Alternatively, state the number of feature values that exceed `n_values_`

arjunjauhari replied:
I am going ahead with number of feature values that exceed n_values_. Just a thought, should i also include the row and col indices of those feature values? Or maybe thats too verbose?

jnothman replied:
Too verbose, IMO. Sometimes it's helpful to give an example, but I think this is straightforward enough that the user can work it out

jnothman reviewed (commented):
Otherwise LGTM, thanks!!

arjunjauhari committed:
Minor change in mask calculation.
Removing fit_transform for UnaryEncoder and relying on one defined in TransformerMixin
Update handle_greater=error error message

Removing fit_transform for UnaryEncoder and relying on one defined in TransformerMixin

Update handle_greater=error error message

jnothman commented:
I now wonder if we should have a default handle_greater='warn'. I think in
practice that raising an error when a count-valued feature exceeds its
training set range is too intrusive. Better off clipping but warning unless
the user has specified otherwise.

arjunjauhari commented:
I like that idea. Its easy to add warn mode, its same as clip mode just with additional warning message. I will add it and make it default. BTW, what's the best way to throw a warning message? a print statement or is there some class which i can use?

jnothman commented:
We usually use warnings.warn

Btw, despite having said LGTM, this would benefit from an example. I'd sort
of imagined that this would be better exemplified by integration with a
discretizer. But if you can think of an example that would make use of this
in our example gallery, feel free to go ahead and propose and add one.

arjunjauhari committed:
Adding warn as a new option for handle_greater parameter.

Updaing warn mode

Making warn as default mode

arjunjauhari committed:
Updating test case to take care of new handle_greater='warn' as default
and updating docs

Updating docs

arjunjauhari commented:
I have added handle_greater='warn' as default mode, but I am a little skeptical about it. Currently, it generates warning when `transform` or `fit_transform` is called. Otherwise, its behavior is same as `clip` option. User can suppress the warning by setting `handle_greater` to `clip`. It almost feel like we can achieve same thing by adding the warning in `clip` option and making it default option. only that user won't have control over suppressing it.

Also, I will look into the example gallery to find something suitable.

jnothman commented:
well the user still have control with the warnings module's filters (which
can hide a warning or turn it into an exception). but I think users are
much more comfortable controlling these things with a visible parameter, in
part because it makes them aware that they can control it.

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,212 @@ def inverse_transform(self, X):
+
+    handle_greater : str, 'warn' or 'error' or 'clip', default='warn'
+        Whether to raise an error or clip or warn if an
+        ordinal feature >= n_values is passed in.
```
clip could do with a description, and warn can be described as clipping with a warning

jnothman commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,212 @@ def inverse_transform(self, X):
+            elif self.handle_greater == 'error':
+                raise ValueError("Found feature values %s which exceeds "
+                                 "n_values during transform."
+                                 % X.ravel()[mask])
```
Not reporting how many?

jnothman reviewed (commented):
The warning needs tests

arjunjauhari committed:
Fixing concerns.
Explaining handle_greater options

Adding test cases to test warn option of handle_greater parameter

Updated warning message

updated feature_matrix generation function and made test deterministic

making test cases clearer by using n_features

n_values and n_features cleanup

TomDLT commented on sklearn/preprocessing/data.py:
```diff
@@ -2873,3 +2876,212 @@ def inverse_transform(self, X):
+                self.n_values == 'auto'):
+            n_values = np.max(X, axis=0) + 1
+        elif isinstance(self.n_values, numbers.Integral):
+            if (np.max(X, axis=0) >= self.n_values).any():
```
Isn't it redundant with the check at this end of `_fit`?
which is better since it uses `handle_greater`

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,218 @@ def test_quantile_transform_valid_axis():
+
+def _generate_random_features_matrix(n_values=3, size=10):
+    rng = np.random.RandomState(0)
+    X = rng.randint(n_values, size=(size, n_values))
```
Is there a reason why `n_features == n_values` ?
Using `n_samples` and `n_features` variable names would be clearer.

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,218 @@ def test_quantile_transform_valid_axis():
+
+def test_unary_encoder_stack():
+    # multiple input features stack to same output
+    n_values = np.random.randint(2, 10)
```
We tend to prefer deterministic tests, using `rng = np.random.RandomState(0)` and `rng.randint` instead of `np.random.randint`.

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,218 @@ def test_quantile_transform_valid_axis():
+    encoder.fit(X)
+
+    # test that an error is raised when different shape
+    larger_n_values = n_values + delta
```
This is not clear.
It should be larger `n_features` and not larger `n_values`.
This is linked to the use of `n_values` as both `n_values` and `n_features` in `_generate_random_features_matrix`.

TomDLT commented on sklearn/preprocessing/tests/test_data.py:
```diff
@@ -2224,3 +2225,218 @@ def test_quantile_transform_valid_axis():
+
+    X = _generate_random_features_matrix(n_values, size)
+    X_trans = enc.fit_transform(X)
+    assert_equal(X_trans.shape, (size, unary_n_values * len(X[0])))
```
`n_features` is clearer than `len(X[0])`

amueller commented:
conflicts and test failures...

arjunjauhari committed:
Resolved conflicts with master

amueller commented:
Haven't really looked at this but not sure if I'm sold on ``UnaryEncoder``. I feel like it's hard to discover. ``OrdinalEncoder`` says what it does, but not how it does it. I guess the user guide helps with discoverability, but I feel a discoverable name would be good, too. Do we have any references for either name? Btw, here's the "unary" wikipedia disambiguation: https://en.wikipedia.org/wiki/Unary I would naturally associate it with 2 and 3, while here we're using 1 and 4.

jnothman commented:
here we're using it in the first sense, yes, which I think is not uncommon,
but is less familiar in computer science and algebra.

To clarify (using the framework of semantic roles), "one hot" in
OneHotEncoder describes the manner of encoding, while "categorical" in
CategoricalEncoder describes the kind of operand. "Unary" would describe
the manner of encoding; "ordinal" the operand. I would expect an
OrdinalEncoder to give me an option to encode strings or numbers as one-hot
or unary or as cardinal integers. We could do that, but I also believe that
the UnaryEncoder could/should support non-ordinal, real-valued input.

amueller commented:
What would ``UnaryEncoder`` to on real-valued input? Discretization?
The familiarity is definitely field dependent and a bit hard to argue. I didn't know what `UnaryEncoder`` would do from the name. As I said, I think finding references might help. This seems like a method from the stats literature, but I couldn't find a good name with a quick google.

jnothman commented:
on non-negative real values it would do the same: x > k for column k of
output

arjunjauhari commented:
@amueller @jnothman Did we converge on the name?

jnothman commented:
I think @amueller has disappeared into teaching land, but others are welcome to state their opinion, including you...

jorisvandenbossche commented on sklearn/preprocessing/data.py:
```diff
@@ -3147,3 +3150,216 @@ def inverse_transform(self, X):
+        Whether to raise an error or clip or warn if an
+        ordinal feature >= n_values is passed in.
+
+        - 'warn' (default): same as clip but with warning.
```
For CategoricalEncoder the default is `error` (and for OneHotEncoder as well). Is there a good reason to deviate in this case?

jorisvandenbossche replied:
I see the discussion about this, @jnotham saying:

> I now wonder if we should have a default handle_greater='warn'. I think in practice that raising an error when a count-valued feature exceeds its training set range is too intrusive. Better off clipping but warning unless the user has specified otherwise.

I understand that reasoning for count-like values. But the examples in the documentation are not count-like, and for those this makes less sense I think as a default. 
(but don't know with which type of data this will be used more often)

jorisvandenbossche commented on sklearn/preprocessing/data.py:
```diff
@@ -3147,3 +3150,216 @@ def inverse_transform(self, X):
+            if self.handle_greater == 'error':
+                raise ValueError("handle_greater='error' but found %d feature"
+                                 " values which exceeds n_values."
+                                 % np.count_nonzero(mask))
```
In principle this check is not needed in case of `'auto'` (but not sure whether a small performance improvement for this case is worth moving it elsewhere)

jorisvandenbossche commented on sklearn/preprocessing/data.py:
```diff
@@ -3147,3 +3150,216 @@ def inverse_transform(self, X):
+        data = np.ones(X_ceil.ravel().sum())
+        out = sparse.coo_matrix((data, (row_indices, column_indices)),
+                                shape=(n_samples, indices[-1]),
+                                dtype=self.dtype).tocsr()
```
Maybe not too important, but it is also possible to directly create a csr matrix instead of first constructing coo and then converting (for CategoricalEncoder I edited the construction of the indices to do this: https://github.com/scikit-learn/scikit-learn/pull/9151/commits/85cf315f0847d09f0f9c21180a1a9cc3cdcc3570)

jorisvandenbossche reviewed (commented):
I added some comments (mainly from the viewpoint of just having implemented CategoricalEncodoer)

I think you still need to add a whatsnew notice?

jnothman commented:
@arjunjauhari, will you be following up @jorisvandenbossche's comments?

arjunjauhari commented:
@jnothman, will address the comments soon. @jorisvandenbossche thanks for the review.

jnothman commented:
@arjunjauhari I would love to have this available for use in the KBinsDiscretizer. Let us know if you would like someone else to finish it

arjunjauhari commented:
@jnothman I apologize for delay. I think its pretty close to finish line. I will finish it up in a week or two.

jnothman commented:
Thanks @arjunjauhari. If can be done this weekend, it has a good chance of being merged next week at development sprints, and merged for release.

It would be great if in or subsequent to this PR, encoding='unary' could be added to the new `KBinsDiscretizer`.

qinhanmin2014 commented:
FYI (Sorry if duplicate with existing comments):
(1) We need to remove ``ordinal_features`` to keep consistent with ``OneHotEncoder``&``KBinsDiscretizer``. Maybe also a note to tell users to use ``ColumnTransfomer``  if they only want to preprocess part of the features.
(2) I think we need an example to demonstrate the usage of it and maybe several references since different people may have different names for this feature.

NicolasHug commented:
@arjunjauhari do you want to finish this soonish?
Else I'll pick it up.

NicolasHug commented:
@arjunjauhari FYI have just opened #12893

cmarmo commented:
@NicolasHug, @jnothman, this PR was superseded by #12893, now closed, as no consensus was reached about the implementation. Is the 'UnaryEncoder' issue still relevant? Is this implementation still useful as a starting point? Or should them be closed? Thanks for your help.

NicolasHug commented:
I think we can close this one too. If the idea of the `UnaryEncoder` is to be resuscitated, I think it should start from #12893 which is more advanced
